plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,464 @@
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	CloseContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = g_steal_pointer (&self->mbim_device),
+		.timeout_id = 0,
+	};
+
+	mbim_device_close (ctx.mbim_device, 5, NULL, fu_mbim_qdu_updater_mbim_device_close_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* we should always have both device cleared, and optionally error set */
+	g_warn_if_fail (ctx.mbim_device == NULL);
+
+	if (ctx.error != NULL) {
+		g_propagate_error (error, ctx.error);
+		return FALSE;
+	}
+
+	g_warn_if_fail (ctx.timeout_id == 0);
+	ctx.timeout_id = g_timeout_add_seconds (15, fu_mbim_qdu_updater_reprobe_timeout, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* update detach right after this */
+	return TRUE;
+}
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+  	GCancellable    *cancellable;
+	GError		*error;
+	guint		 timeout_id;
+	GBytes		*blob;
+	GArray		*digest;
+	GPtrArray	*chunks;
+	guint 		chunk_sent;
+	FuDevice 	*device;
+} WriteContext;
+
+static gboolean
+fu_mbim_qdu_updater_reboot_timeout (gpointer user_data)
+{
+	WriteContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_mbim_qdu_updater_file_write_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	g_autoptr(MbimMessage) request  = NULL;
+	g_autoptr(MbimMessage) response = NULL;
+	g_autoptr(GError)      error = NULL;
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_file_write_response_parse (response, &error)) {
+		g_debug ("error: couldn't parse response message: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	fu_device_set_progress_full (FU_DEVICE (ctx->device), (gsize) ctx->chunk_sent, (gsize) ctx->chunks->len);
+	ctx->chunk_sent++;
+	if (ctx->chunk_sent < ctx->chunks->len) {
+		FuChunk *chk = g_ptr_array_index (ctx->chunks, ctx->chunk_sent);
+		request = mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), NULL);
+		mbim_device_command (ctx->mbim_device,
+				     request,
+				     20,
+				     ctx->cancellable,
+				     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+				     ctx);
+		mbim_message_unref (request);
+	} else {
+		fu_device_set_progress (FU_DEVICE (ctx->device), 100);
+		g_debug ("Done!");
+		fu_device_set_status (FU_DEVICE (ctx->device), FWUPD_STATUS_DEVICE_RESTART);
+		g_warn_if_fail (ctx->timeout_id == 0);
+		/* device will auto reboot right after update finish */
+		ctx->timeout_id = g_timeout_add_seconds (30, fu_mbim_qdu_updater_reboot_timeout, ctx);
```

Comment:
I'm a bit worried about these arbitrary 30s here. What happens during those 30s? Does the MBIM port go away completely from the kernel, and then come back? or does the MBIM port "stay" but we cannot use it during those 30s?

Reply:
Adding this kind of 30s timeout, or the other 15s timeout in the same file, is a big broken if you ask me. It's a bit arbitrary, isn't it? What if a future upgrade will need slightly more time due to some reason we don't know of? E.g. what if a future upgrade needs to trigger a new internal filesystem check that adds some 30s more to the process?

Instead of setting up a timeout and just assuming the process is over after that timeout, I think that a safer approach would be to setup a much much shorter timeout; e.g. 5s; then check if the process has finished already (e.g. is the MBIM port reporting the new requested fw version already?) and if the result is no, then wait 5s more, and recheck. And so on. And then add a "max time" threshold; e.g. if we repeat the checks every 5s and we didn't get success after 40 attempts (200s?) then we consider the upgrade failed. Would this kind of logic work?

The key is to know what happens when trying to access the MBIM port during that upgrade time. The MBIM port is there; so what happens if you send a request? The request is not served? The modem crashes?

Reply:
>  What if a future upgrade will need slightly more time due to some reason we don't know of? E.g. what if a future upgrade needs to trigger a new internal filesystem check that adds some 30s more to the process?

OK, I think the situation you described should be considered, the fixed timeout (30s and 15s) is too special, it shouldn't be used, and need to be removed.

> Would this kind of logic work?

Sure, It seems feasible. 
But now, the logic is just wait 120s to re-probe modem after modem uninhibited, is my understanding right? 
```
/* Amount of time for the modem to be re-probed and exposed in MM after being
 * uninhibited. The timeout is long enough to cover the worst case, where the
 * modem boots without SIM card inserted (and therefore the initialization
 * may be very slow) and also where carrier config switching is explicitly
 * required (e.g. if switching from the default (DF) to generic (GC).*/
#define FU_MM_DEVICE_REMOVE_DELAY_REPROBE	120000	/* ms */
```

> The key is to know what happens when trying to access the MBIM port during that upgrade time. The MBIM port is there; so what happens if you send a request? The request is not served? The modem crashes?

Maybe this problem need to consider different modules. 
If the module has more than one partition, and the download firmware is a free partition, it will not affect access the MBIM port during upgrade time.
If the module has only one partition, it will be affected when access the MBIM port during upgrade time, and this access should not be done in this case .

Reply:
> If the module has only one partition, it will be affected when access the MBIM port during upgrade time, and this access should not be done in this case .

This is the thing. In your module, what happens if you try to run MBIM requests *during* the upgrade? I would assume the requests would just timeout. If we poll every once in a while; e.g. request, timeout, wait some seconds, request, timeout, wait some seconds... at some point the request will not timeout and we will get a response. Can't we use this logic to know when the module is again ready after the upgrade?

Reply:
> In your module, what happens if you try to run MBIM requests _during_ the upgrade? 

In our module, there is an extra OTA partition for downloading firmware, so the module can access the Internet normally during the process of downloading firmware. After downloading, the module will restart,  and we can not run any MBIM requests during the restart process.
>  Can't we use this logic to know when the module is again ready after the upgrade?

Sure, we can. I think this logic should start after the firmware is downloaded.

Reply:
> After downloading, the module will restart, and we can not run any MBIM requests during the restart process.

What happens if you try to run a MBIM command *during* the restart process? This question really only applies because the MBIM port is kept exposed during the whole process. Will the commands just timeout?

I think that is more or less what we should attempt. E.g.:
 * Download firmware file via MBIM
 * Once the download is finished, we can wait some time, e.g. 30s as you originally had.
 * Then, we try to run the MBIM command that queries firmware version.
 * If the command is successful, we check the reported firmware version. If the firmware version is the one we want, we're done. If the firmware version is not the one we want, we report a failure?
 * If the command is not successful (e.g. timed out because the modem is not serving MBIM operations), then we schedule a new check with a timeout in e.g. 10s. After those 10s we retry the MBIM command that queries firmware version.
 * And we repeat the process; if the timeout happens again, we reschedule after 10s.
 * We could have a max threshold of e.g. 10 or 15 retries (e.g. 100s or 150s in addition to the initial 30s).

What do you think?

It's fine to have the 30s timeout, but I would say that only if we don't just assume everything is fine after those 30s. We need to check explicitly, and if it's not right, we reschedule the check.

If we don't do this and the 30s is too tight, it may happen that you attempt to access the modem while it's not yet ready to serve MBIM operations. And that may be exactly why you had that mbim-proxy kill operation, or why you required the pci rescan, so that the MBIM port re-probing logic performs those timed checks on the port. We should do those checks ourselves in fwupd, I'd say, and only assume the upgrade is done once the MBIM port is responsive again (as seen by fwupd).

Reply:
Also, in this suggested process, there should be no explicit update_detach/update_attach procedure needed, as the modem doesn't change port layout.

Reply:
> What happens if you try to run a MBIM command during the restart process? This question really only applies because the MBIM port is kept exposed during the whole process. Will the commands just timeout?
I think that is more or less what we should attempt. 

OK, I will try it.

From your suggestions, I think I need two more works to make this new upgrade method more right.
1. No explicit update_detach/update_attach procedure needed.
2. When firmware download is finished, wait some time, then add a retry action as you said above.

Is my understanding correct? Is there anything else missing?

Reply:
Yes, those are my suggestions